### ISSUE TITLE
Add a retry logic to the verifyHTTPResponse() step

### ISSUE THREAD

**Issue by agajdosi** [minishift/minishift#1816]
This is upstream follow up issue for the [CDK-222](https://issues.jboss.org/browse/CDK-222). In combination of slower host and more complicated application on OpenShift, the time for the http endpoint of the application to be accessible is longer and thus some tests may fail.

To fix this behaviour and secure that the result reported from tests is relevant and true, we should add a retry logic to this step, so when it takes more time for app to be ready, the tests will wait and try it again in reasonable time.

**Comment by gbraad** [comment]
I do not think a retry logic is correct. If a 500 or 503 is returned, the service failed. Bashing it another time is just not a proper test. I'd rather see an explicit backing off (sleep) as step or part of a deployment... WDYT?

**Comment by agajdosi** [comment]
@gbraad I was thinking about sleep also, but how much we want to have? 30 seconds? In most cases the tests passes just fine as they are now, on some machines maybe just 3 or 5 seconds would help. On the machine I am facing a problem 5 second sleep helps. So I understand the retry as more sophisticated sleep in this case.

But well, I understand also that having the sleep mentioned directly in the Gherkin feature maybe is what we want to have, to be transparent and understandable.

**Comment by hferentschik** [comment]
Sleeping is never a good option. Does this demo use a readiness probe? If so, should the rollout not wait for the readiness probe to report success. In this case it should be save to hit the endpoint. 

I think before adding random sleeps we should investigate what's going on here.

**Comment by gbraad** [comment]
it is not a 'sleep' that needs to happen, but an ease-off. It seems the service is slow to setup correctly. So when roll-out succeeded, the service is not available yet?

retriggering the service seems wrong... so, what about a health check before checking the services or as suggested maybe a readiness-probe? Although, these might lead to the same result I guess.

**Comment by agajdosi** [comment]
> retriggering the service seems wrong (which was suggested before)

No @gbraad, I have never suggested to add a redeployment of the service/app. I have suggested to add a retry logic to the the verifyHTTPResponse() which is the step that checks the http response. If the response is wrong, the step should wait a little bit and try once more.

> Does this demo use a readiness probe? If so, should the rollout not wait for the readiness probe to report success.

@hferentschik Did not check yet. My first plan was to add a retry logic to http response check, so one does not need to care if the application she wants to test use the readiness probes or not. But I agree, the readiness probe check could be part of the rollout step - if it is possible to handle the applications which do not use the probe.

**Comment by gbraad** [comment]
retriggering, in the sense of touch it again. not a redeploy ;-) (retry logic as the title suggest) <= I do not think this is what we should do.

**Comment by gbraad** [comment]
I am curious about the readiness probe, as after the rollout the application is actually 'ready', but seems to still have a '500'/'503' due to being unavailable. It is worth a try.

**Comment by LalatenduMohanty** [comment]
> To fix this behaviour and secure that the result reported from tests is relevant and true, we should add a retry logic to this step

I would prefer a retry logic as sleep is very vague in terms of how much time it takes in case the service is slow to come online. I think this will also help #1808

**Comment by gbraad** [comment]
as discussed, I think we should try the readiness probe first (as we can create them from the command line), and if that does not work we can do an alternative.

**Comment by agajdosi** [comment]
@gbraad The addition of readiness probe does not fix the issue:

```gherkin
14:11:48     Examples: Required information to test the templates
14:11:48       | template-name    | service-name | http-endpoint | expected-hello                        |
14:11:48       | datagrid65-basic | datagrid-app | /             | Welcome to the JBoss Data Grid Server |
14:11:48         body of reponse from http://datagrid-app-myproject.192.168.42.5.nip.io does not contain expected string. Expected: Welcome to the JBoss Data Grid Server, Actual: <html>
```

Full log at https://ci.centos.org/job/minishift-pr/3038/consoleFull.

Looking for possible alternatives to `rollout` command and will test them, if there are any.

**Comment by gbraad** [comment]
How about the health check. It can be configured with Liveliness and Readiness taking an `InitialDelayInSeconds`. this can help in taking the sleep/delay/ease-off out of our actual scenario

**Comment by agajdosi** [comment]
@gbraad This can be done, but I am somehow asking myself how is:

```
oc set probe dc/ruby-ex --readiness --get-url=http://:8080 --initial-delay-seconds=5
```

different from just sleeping directly in tests? Also it feels that we are doing something the developers of the examples should do - having a health check and maybe initial delay tuned for the needs of the particular application.

For now I would say to go for a retry of http request implemented in the internal logic of the step definition. Looks like good choice to me as it saves time and mimics user's behaviour.

**Comment by gbraad** [comment]
This is also my take on this.

I have no real problem with a sleep, as long as it is explicit. In real
life a user can also not bash the buttons on his keyboard and mouse that
quickly.

However, I'd still rather see it happen from a probe.

On Thursday, December 14, 2017, Andreas Gajdosik <notifications@github.com>
wrote:
> @gbraad This can be done, but I am somehow asking myself how is:
>
> oc set probe dc/ruby-ex --readiness --get-url=http://:8080
--initial-delay-seconds=5
>
> different from just sleeping directly in tests? Also it feels that we are
doing something the developers of the examples should do - having a health
check and maybe initial delay tuned for the needs of the particular
application.
>
> For now I would say to go for a retry of http request implemented in the
internal logic of the step definition. Looks like good choice to me as it
saves time and mimics user's behaviour.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci5.googleusercontent.com/proxy/dVRvwxOMttrbWVjzAxM5RizHzIKqcNUEjd3rXhb6EgKu2m9orbAoMeFEjGYKiE9D9b0ba0C07R382zQ77wkBryXTVL_-Dl0U52t2ZIEVEUKv8L9e-YrpOKCYCoU9lo_XsymBHcUDgN-H3e7eYb4ThIYDYzjIQg=s0-d-e1-ft#https://github.com/notifications/beacon/AAAHZpKt8XusvyJhWkAQao9DSsMM-iE3ks5tAOzggaJpZM4Q5p03.gif
>

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by LalatenduMohanty** [comment]
Resolved via #1838
